text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,303 @@
+
+### User Interaction
+
+Compile time features for `core` may be specified using the same Cargo.toml syntax used for other crates.
+
+The syntax is the same when using `unstable` and `stable` features, however the former may only be used with a nightly compiler, and use of an `unstable` feature with a stable compiler would result in a compile time error.
+
+The syntax for these features would look as follows:
+
+```toml
+[dependencies.core]
+default-features = false
+features = [...]
+```
+
+It is not necessary to explicitly mentioned the dependency of `core`, unless changes to features are necessary.
+
+### Technical Implications
+
+#### Path to stabilization
+
+The stabilization of a `core` feature flag would require a process similar to the stabilization of a feature in the language:
+
+* Any new feature begins as unstable, requiring a nightly compiler
+* When the feature is sufficiently explored, an RFC/PR can be made to `libcore` to promote this feature to stable
+* When this has been accepted, the feature of `core` may be used with the stable compiler.
+
+#### Implementation of Stable Features
+
+There would be some mechanism of differentiating between flags used to build core, sorting them into the groups `unstable` and `stable`. This RFC does not prescribe a certain way of implementation.
+
+## 3 - Extend the new behaviors described for `std` (and `alloc`)
+
+### Use Case
+
+Once the design and implications of the changes have been made for `core`, it will be necessary to extend these abilities for `std`, including components like `liballoc`.
+
+### Caveats
+
+In general, the same restrictions for building `core` will apply to building `std`. These include:
+
+* Users of the stable compiler must use the source used to build the current rust compiler
+* Only compile time features considered `stable` may be used outside of nightly. Initially the list of `stable` features would be empty, and stabilizing these features would require a PR/RFC to `libstd`.
+
+### User Interaction
+
+The building of `std` would respect the current build profile, including
+
+The syntax for these features would look as follows:
+
+```toml
+[dependencies.std]
+default-features = false
+features = [
+    "force_alloc_system",
+]
+```
+
+It is not necessary to explicitly mention the dependency of `std`, unless changes to features are necessary.
+
+### Technical Implications
+
+None beyond the technical implications listed for `core`.
+
+## 4 - Allow the user to provide their own custom source versions of `core` and `std`
+
+### Use Case
+
+This will allow users of a nightly compiler to provide a custom version of `core` and `std`, without requiring the recompilation of the compiler itself.
+
+### Caveats
+
+As stability guarantees cannot be made around modified versions of `core` or `std`, a nightly compiler must always be used.
```

Comment:
Why replacing `core` or `std` with a custom version requires nightly? How is it different from `no_std` and cargo source replacement which are both available on stable and allow similar effects? It makes `core` and `std` special again because the user has the freedom to use custom source for any other crate but for some reason is not allowed to use custom source for `core` or `std`.

Reply:
It seems to me that just replacing `core` or `std` with a custom version should be fine on stable, those custom versions should not be allowed to bypass stability like the standard version, so they will almost certainly require nightly to compile, but that's separate from the act of replacement.